### PR TITLE
[Fix](Nereids) Fix bitmap type in compare operator bug

### PR DESCRIPTION
**Pull request by LiBinfeng-01**
# Proposed changes

## Problem summary

When using nereids, if we use compare operator of bitmap type, an analyze exception need to be throwed.
like: select id from (select BITMAP_EMPTY() as c0 from expr_test) as ref0 where c0 = 1 order by id
Which c0 in subq0 is a bitmap type, this scenario is not supported right now.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by LiBinfeng-01** [comment]
run buildall

**Comment by LiBinfeng-01** [comment]
run buildall

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java [hunk 1]
```diff
@@ -58,4 +59,16 @@ public AbstractDataType inputType() {
      * Commute between left and right children.
      */
     public abstract ComparisonPredicate commute();
+
+    @Override
+    public TypeCheckResult checkInputDataTypes() {
+        for (int i = 0; i < super.children().size(); i++) {
+            Expression input = super.children().get(i);
+            if (input.getDataType().isBitmapType()) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java:67]: use `isObjectType()` to ban all object types

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java [hunk 2]
```diff
@@ -119,6 +120,12 @@ private boolean childrenAtLeastOneInOrExistsSub(BinaryOperator binaryOperator) {
                 || binaryOperator.right().anyMatch(Exists.class::isInstance);
     }
 
+    private boolean childrenAtLeastOneIsBitMapType(Expression binaryOperator) {
+        return binaryOperator.getArgument(0).getDataType() instanceof BitmapType
+
+            || binaryOperator.getArgument(1).getDataType() instanceof BitmapType;
+    }
+
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java:128]: remove useless code

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java (hunk 1 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/trees/expressions/ComparisonPredicate.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java (hunk 2 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/SubExprAnalyzer.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java [hunk 3]
```diff
@@ -46,6 +51,13 @@ public List<Rule> buildRules() {
                     checkBound(plan);
                     return null;
                 })
+            ),
+            RuleType.CHECK_BITMAP_ANALYSIS.build(
+                logicalHaving().thenApply(ctx -> {
+                    LogicalHaving<Plan> having = ctx.root;
+                    checkHavingBitmapExpression(having);
+                    return null;
+                })
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java:60]: add comment about this rule to explain why we need this check and what will happen if we do not have this rule

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java [hunk 4]
```diff
@@ -71,4 +83,18 @@ private void checkBound(Plan plan) {
             ));
         }
     }
+
+    private void checkHavingBitmapExpression(LogicalHaving<Plan> having) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java:87]: rename Bitmap to ObjectType

File: regression-test/suites/nereids_arith_p0/bitmap.groovy [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+suite('nereids_arith_p0_date') {
+    sql 'use regression_test_nereids_arith_p0'
+    sql 'set enable_nereids_planner=true'
+    sql 'set enable_fallback_to_original_planner=false'
+    test {
+        sql "select id from (select BITMAP_EMPTY() as c0 from expr_test) as ref0 where c0 = 1 order by id"
+        exception "errCode = 2, detailMessage = Unexpected exception: can not cast from origin type BITMAP to target type=DOUBLE"
```
Comment (morrySnow) [inline regression-test/suites/nereids_arith_p0/bitmap.groovy:24]: this exception msg is not match ur new check in ComparisonPredicate. do we really need this check? or ur regression test do not cover the case u want to test?

Comment (LiBinfeng-01) [reply regression-test/suites/nereids_arith_p0/bitmap.groovy:24]: done, I fix two scenario, so I add two cases

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java (hunk 4 above)
Comment (LiBinfeng-01) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/CheckBound.java]: done

**Comment by LiBinfeng-01** [comment]
run p0

File: fe/fe-core/src/main/java/org/apache/doris/nereids/fe.pid
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/fe.pid:1]: remove this file

**Comment by LiBinfeng-01** [comment]
run buildall

**Review by morrySnow — APPROVED** [review]

**Pull request merged**